### PR TITLE
Serializer setting

### PR DESCRIPTION
**Pull request by lisemon**
Hmm, how about something like this? 😄

### CODE REVIEW

**Comment by RSuter** [comment]
Ref https://github.com/RSuter/NJsonSchema/issues/495

File: src/NJsonSchema/JsonSerializerSettings.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+        override
+        public string ToString() {
+            var settingsString = DateParseHandling == null ? 
+                                    string.Empty : 
+                                    $"DateParseHandling = {DateParseHandling.ToString()}";
+            
+            settingsString += string.IsNullOrEmpty(DateFormatString) ? 
+                                    string.Empty : 
+                                    $"DateFormatString = {DateFormatString}";
```
Comment (RSuter) [inline src/NJsonSchema/JsonSerializerSettings.cs]: I think we need to escape here with `\"`, eg

      $"DateFormatString = \"{DateFormatString}\"";

File: src/NJsonSchema/JsonSerializerSettings.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public Newtonsoft.Json.DateParseHandling? DateParseHandling { get; set; }
+
+        public string DateFormatString { get; set; }
+
+        override
+        public string ToString() {
+            var settingsString = DateParseHandling == null ? 
+                                    string.Empty : 
+                                    $"DateParseHandling = {DateParseHandling.ToString()}";
```
Comment (RSuter) [inline src/NJsonSchema/JsonSerializerSettings.cs]: I think the namespace and type is missing, eg. 

    $"DateParseHandling = Newtonsoft.Json.DateParseHandling.{DateParseHandling.ToString()}";

Comment (lisemon) [reply src/NJsonSchema/JsonSerializerSettings.cs]: It is referring to the Property DateParseHandling inside JsonSerializerSettings.

**Comment by lisemon** [comment]
@RSuter Thanks for the quick response 👍  Fixed the stuff.

**Comment by RSuter** [comment]
I have to check the implications of this in the NSwag project before i can merge it... thanks for the pr

**Comment by RSuter** [comment]
I think I'll move the two settings to CSharpGeneratorSettings - to avoid having a collision with the JsonSerializerSettings class of Newtonsoft.Json and because CSharpGeneratorSettings already contains serializer relevant settings... (most of them). Ok for you?

**Comment by lisemon** [comment]
The reason I put the two settings in a separate class is because there are a quite a bunch of serializer settings, that potentially could supported in the long run. But for now I only need the two date specific settings. And as it is not really a generator setting, but more a bundle of serializing setting.
Please have that in mind - but do as you like.

**Comment by RSuter** [comment]
Another completely different option would be to add the setting "JsonSerializerSettingsFactoryMethod" (e.g. "MyNamespace.MyClass.MyStaticMethod" which - when set - gets called in `GenerateJsonSerializerParameterCode()` and you have to implement it somewhere in the project. This way we would be much more flexible and you can configure everything (even converters, switch the contract resolver, etc.). What do you think? This would be a single setting  in CSharpGeneratorSettings...

**Comment by lisemon** [comment]
It is nice to be able to configure a lot 😄  As long as it is well-documented.

**Comment by RSuter** [comment]
Yes, i like the "JsonSerializerSettingsFactoryMethod" setting more, its just one setting which is very flexible, also i only need to add just one new setting to the command line and the UI (see http://nswag.org)

**Comment by lisemon** [comment]
Sounds good 👍

**Comment by RSuter** [comment]
Can you change this? Or should I do it? Id like to release today if possible :-)

**Comment by lisemon** [comment]
Can you do it? I won't be able to look at it until tomorrow (and I'm in a +02:00 timezone)

**Comment by RSuter** [comment]
Ok no problem, i think ill call the setting "JsonSerializerSettingsTransformationMethod" with the signature: 

    JsonSerializerSettings Transform(JsonSerializerSettings settings)

This way we can provide the default settings which can be transformed or completely replaced... sound good?

**Comment by RSuter** [comment]
Released in v9.7, we can still tweak it as long as no one uses it :-)

**Comment by lisemon** [comment]
Awesome with the quick iteration and release!!

I seems to compile fine for serializing, but not for deserializing, as the deserializer does not have the Formatting parameter.

https://www.newtonsoft.com/json/help/html/SerializeObject.htm
https://www.newtonsoft.com/json/help/html/DeserializeObject.htm

E.g.
`csharpGenerator.Settings.JsonSerializerSettingsTransformationMethod = "SerializerSettings.SetupSerializer";`

```
public class SerializerSettings
    {
        public static Formatting SetupSerializer(JsonSerializerSettings jsonSerializerSettings)
        {
            jsonSerializerSettings.DateParseHandling = DateParseHandling.DateTimeOffset;
            return jsonSerializerSettings.Formatting;
        }
    }
}
```

```
public static Visualization FromJson(string data)
        {
            return Newtonsoft.Json.JsonConvert.DeserializeObject<Visualization>(data, SerializerSettings.SetupSerializer(new Newtonsoft.Json.JsonSerializerSettings {  }));
        }

 public string ToJson() 
        {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this, SerializerSettings.SetupSerializer(new Newtonsoft.Json.JsonSerializerSettings {  }));
        }
```

Am I getting something wrong?

**Comment by lisemon** [comment]
Furthermore the setting seems not to be used in the `JsonInheritanceConverter`.

**Comment by RSuter** [comment]
The signatures are: 

    public static T DeserializeObject<T>(string value, JsonSerializerSettings settings)
    public static string SerializeObject(object value, JsonSerializerSettings settings)

**=> Ah sorry, just saw - `SetupSerializer` must return `JsonSerializerSettings`.**

JsonInheritanceConverter: 

Read or write?
For reading we are using the serializer (with hopefully the correct settings): 
https://github.com/RSuter/NJsonSchema/blob/master/src/NJsonSchema/Converters/JsonInheritanceConverter.cs#L120

**Comment by RSuter** [comment]
If you change the SetupSerializer signature, maybe the JsonInheritanceConverter also works correctly?

**Comment by RSuter** [comment]
I've updated the wiki: https://github.com/RSuter/NJsonSchema/wiki/CSharpGeneratorSettings

**Comment by lisemon** [comment]
Ahh yes! 😄 Sorry about that!

I will have to do some testing in order to confirm that it works in the InheritanceConverter 😄

**Comment by RSuter** [comment]
If possible, create a PR with a unit test :-)

**Comment by RSuter** [comment]
... so that it will also not break in the future

**Comment by lisemon** [comment]
Awesome! Maan, you're fast 😄

**Comment by lisemon** [comment]
So at last, I had time to go through with testing... The InheritanceConverter does not use the setting in neither in `ReadJson` or `WriteJson`.

I am mostly concerned with the ReadJson. It is actually the JsonReader that needs the setting. I will have a look at this for the next hour,  and possibly write a unittest.

**Comment by RSuter** [comment]
This is bad... and strange: The InheritanceConverter uses the serializer in both read and write so it should use the given settings...

You can write tests against

https://github.com/RSuter/NJsonSchema/blob/master/src/NJsonSchema/Converters/JsonInheritanceConverter.cs

If everything works, I'll copy this class into the template which is used to generate code.

**Comment by RSuter** [comment]
And we should open a new issue for that...

**Comment by lisemon** [comment]
👍  Jup!

**Comment by lisemon** [comment]
Oh boy I am not even close to having something to push! :D

**Comment by lisemon** [comment]
So I cannot run the test project due to environent issues; as the project type for NJsonSchema.CodeGeneration.Tests is not supported for Visual Studio Community for Mac. I guess it has most todo with the structure of the .csproj-file.

I however note that the `JsonInheritanceConverterTemplateModel` gets the settings in the constructor, and that `JsonInheritanceConverterTemplate.tt` contains the code that is used to generate the .cs-file. I am pretty lost as to how the generation of the .cs-file works but what I can do, it to try and modify the generated model file to make it do as I want, and then we can maybe use that in the .tt-file?

**Comment by RSuter** [comment]
Yes, the test project are not yet .NET Standard - they are legacy MSBuild project using full .NET 4.6. I have to migrate them to .NET Standard and xUnit...

Yes, please change the generated code unit it works and i'll run the tests with it and updated everything...

**Comment by lisemon** [comment]
Thanks. Sorry for not being of more help.

**Comment by RSuter** [comment]
No problem - you are helping... can you try changing the generated code and report here if you've found a solution?

**Comment by lisemon** [comment]
yes!

**Pull request closed without merging**